File: examples_android/app/src/main/java/com/confirmit/testsdkapp/layouts/defaultlayoutjava/DefaultSurveyAdapter.java
package com.confirmit.testsdkapp.layouts.defaultlayoutjava;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.confirmit.testsdkapp.R;
import com.confirmit.testsdkapp.layouts.defaultlayoutjava.cells.DefaultLayoutCell;
import com.confirmit.testsdkapp.layouts.defaultlayoutjava.cells.DefaultLayoutClickLabelCell;
import com.confirmit.testsdkapp.layouts.defaultlayoutjava.cells.DefaultLayoutIconLabelCell;
import com.confirmit.testsdkapp.layouts.defaultlayoutjava.cells.DefaultLayoutLabelCell;
import com.confirmit.testsdkapp.layouts.defaultlayoutjava.cells.DefaultLayoutSeparatorCell;
import com.confirmit.testsdkapp.layouts.defaultlayoutjava.cells.DefaultLayoutTextBoxCell;
import com.confirmit.testsdkapp.layouts.defaultlayoutjava.items.DefaultLayoutClickLabel;
import com.confirmit.testsdkapp.layouts.defaultlayoutjava.items.DefaultLayoutIconLabel;
import com.confirmit.testsdkapp.layouts.defaultlayoutjava.items.DefaultLayoutItem;
import com.confirmit.testsdkapp.layouts.defaultlayoutjava.items.DefaultLayoutLabel;
import com.confirmit.testsdkapp.layouts.defaultlayoutjava.items.DefaultLayoutSection;
import com.confirmit.testsdkapp.layouts.defaultlayoutjava.items.DefaultLayoutTextBox;
import com.confirmit.testsdkapp.layouts.defaultlayoutjava.items.DefaultSurveyCellType;

import java.util.ArrayList;
import java.util.List;

public class DefaultSurveyAdapter extends RecyclerView.Adapter<DefaultLayoutCell> {
    private List<Integer> sectionPositions = new ArrayList<>();
    private List<DefaultSurveySectionInfo> sectionInfos = new ArrayList<>();
    private List<DefaultLayoutSection> sections;

    public DefaultSurveyAdapter(List<DefaultLayoutSection> sections) {
        super();
        this.sections = sections;
    }

    public void refresh() {
        getItemCount();
    }

    public int getItemIndex(int section, int row) {
        return sectionInfos.get(section).getItemAbsoluteIndex(row);
    }

    @Override
    public int getItemViewType(int position) {
        if (sectionPositions.contains(position)) {
            return DefaultSurveyCellType.PAGE_HEADER.intValue();
        } else {
            for (DefaultSurveySectionInfo sectionInfo : sectionInfos) {
                if (!sectionInfo.isInSection(position)) {
                    continue;
                }

                DefaultLayoutSection section = sections.get(sectionInfo.getIndex());
                DefaultLayoutItem item = section.getItems().get(sectionInfo.getItemIndex(position));
                if (item instanceof DefaultLayoutLabel) {
                    return DefaultSurveyCellType.LABEL.intValue();
                }
                if (item instanceof DefaultLayoutIconLabel) {
                    return DefaultSurveyCellType.ICON_LABEL.intValue();
                }
                if (item instanceof DefaultLayoutTextBox) {
                    return DefaultSurveyCellType.TEXTBOX.intValue();
                }
                if (item instanceof DefaultLayoutClickLabel) {
                    return DefaultSurveyCellType.CLICK_LABEL.intValue();
                }
            }
        }

        return DefaultSurveyCellType.PAGE_HEADER.intValue();
    }

    @NonNull
    @Override
    public DefaultLayoutCell onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        DefaultSurveyCellType type = DefaultSurveyCellType.fromValue(viewType);
        View view;
        switch (type) {
            case PAGE_HEADER:
                view = LayoutInflater.from(parent.getContext()).inflate(R.layout.default_layout_section_header, parent, false);
                return new DefaultLayoutSeparatorCell(view);
            case LABEL:
                view = LayoutInflater.from(parent.getContext()).inflate(R.layout.default_layout_label, parent, false);
                return new DefaultLayoutLabelCell(view);
            case ICON_LABEL:
                view = LayoutInflater.from(parent.getContext()).inflate(R.layout.default_layout_icon_label, parent, false);
                return new DefaultLayoutIconLabelCell(view);
            case TEXTBOX:
                view = LayoutInflater.from(parent.getContext()).inflate(R.layout.default_layout_textbox, parent, false);
                return new DefaultLayoutTextBoxCell(view);
            case CLICK_LABEL:
                view = LayoutInflater.from(parent.getContext()).inflate(R.layout.default_layout_click_label, parent, false);
                return new DefaultLayoutClickLabelCell(view);
        }

        return null;
    }

    @Override
    public void onBindViewHolder(@NonNull DefaultLayoutCell holder, int position) {
        for (DefaultSurveySectionInfo sectionInfo : sectionInfos) {
            if (sectionInfo.isInSection(position)) {
                DefaultLayoutSection section = sections.get(sectionInfo.getIndex());
                DefaultLayoutItem item = section.getItems().get(sectionInfo.getItemIndex(position));
                holder.setup(item, section);
            }
        }
    }

    @Override
    public int getItemCount() {
        sectionPositions.clear();
        sectionInfos.clear();

        int count = 0;
        int minRowIndex = 1;
        int maxRowIndex = 0;
        int totalSectionCount = getTotalSectionCount();

        for (int i = 0; i < totalSectionCount; i++) {
            int rowCount = getSectionItemCount(i);
            maxRowIndex = minRowIndex + rowCount;
            sectionPositions.add(i, minRowIndex - 1);
            sectionInfos.add(i, new DefaultSurveySectionInfo(i, minRowIndex, maxRowIndex, rowCount));

            count += rowCount;

            minRowIndex = maxRowIndex + 1;
        }

        return count + totalSectionCount;
    }


    private int getTotalSectionCount() {
        return sections.size();
    }

    private int getSectionItemCount(int section) {
        return sections.get(section).getItems().size();
    }
}

File: examples_android/app/src/main/java/com/confirmit/testsdkapp/layouts/defaultlayout/DefaultSurveyAdapter.kt
package com.confirmit.testsdkapp.layouts.defaultlayout

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.confirmit.testsdkapp.R
import com.confirmit.testsdkapp.layouts.defaultlayout.cells.*
import com.confirmit.testsdkapp.layouts.defaultlayout.items.*

class DefaultSurveySectionInfo(val index: Int, private val minRowIndex: Int, private val maxRowIndex: Int, val rowCount: Int) {
    fun isInSection(position: Int): Boolean {
        return position in minRowIndex..(maxRowIndex - 1)
    }

    fun getItemIndex(position: Int): Int {
        return position - minRowIndex
    }

    fun getItemAbsoluteIndex(row: Int): Int {
        return minRowIndex + row
    }
}

enum class DefaultSurveyCellType(val value: Int) {
    PAGE_HEADER(0),
    LABEL(1),
    ICON_LABEL(2),
    TEXTBOX(3),
    CLICK_LABEL(4);

    companion object {
        fun fromValue(value: Int): DefaultSurveyCellType? = DefaultSurveyCellType.values().find { it.value == value }
    }
}

class DefaultSurveyAdapter(private var sections: MutableList<DefaultLayoutSection>) : RecyclerView.Adapter<DefaultLayoutCell>() {
    private val sectionPositions = mutableListOf<Int>()
    private val sectionInfos = mutableListOf<DefaultSurveySectionInfo>()

    fun refresh() {
        itemCount // Reload section Infos
    }

    fun getItemIndex(section: Int, row: Int): Int {
        return sectionInfos[section].getItemAbsoluteIndex(row)
    }

    override fun getItemViewType(position: Int): Int {
        if (sectionPositions.contains(position)) {
            return DefaultSurveyCellType.PAGE_HEADER.value
        } else {
            sectionInfos.find { it.isInSection(position) }?.let {
                val section = sections[it.index]
                val item = section.items[it.getItemIndex(position)]
                when (item) {
                    is DefaultLayoutLabel -> {
                        return DefaultSurveyCellType.LABEL.value
                    }
                    is DefaultLayoutIconLabel -> {
                        return DefaultSurveyCellType.ICON_LABEL.value
                    }
                    is DefaultLayoutTextBox -> {
                        return DefaultSurveyCellType.TEXTBOX.value
                    }
                    is DefaultLayoutClickLabel -> {
                        return DefaultSurveyCellType.CLICK_LABEL.value
                    }
                    else -> {
                        return DefaultSurveyCellType.PAGE_HEADER.value
                    }
                }
            }
        }

        return DefaultSurveyCellType.PAGE_HEADER.value
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DefaultLayoutCell {
        val type = DefaultSurveyCellType.fromValue(viewType)!!
        when (type) {
            DefaultSurveyCellType.PAGE_HEADER -> {
                val view = LayoutInflater.from(parent.context).inflate(R.layout.default_layout_section_header, parent, false)
                return DefaultLayoutSeparatorCell(view)
            }
            DefaultSurveyCellType.LABEL -> {
                val view = LayoutInflater.from(parent.context).inflate(R.layout.default_layout_label, parent, false)
                return DefaultLayoutLabelCell(view)
            }
            DefaultSurveyCellType.ICON_LABEL -> {
                val view = LayoutInflater.from(parent.context).inflate(R.layout.default_layout_icon_label, parent, false)
                return DefaultLayoutIconLabelCell(view)
            }
            DefaultSurveyCellType.TEXTBOX -> {
                val view = LayoutInflater.from(parent.context).inflate(R.layout.default_layout_textbox, parent, false)
                return DefaultLayoutTextBoxCell(view)
            }
            DefaultSurveyCellType.CLICK_LABEL -> {
                val view = LayoutInflater.from(parent.context).inflate(R.layout.default_layout_click_label, parent, false)
                return DefaultLayoutClickLabelCell(view)
            }
        }
    }

    override fun onBindViewHolder(holder: DefaultLayoutCell, position: Int) {
        sectionInfos.find { it.isInSection(position) }?.let {
            val section = sections[it.index]
            val item = section.items[it.getItemIndex(position)]
            holder.setup(item, section)
        }
    }

    override fun getItemCount(): Int {
        sectionPositions.clear()
        sectionInfos.clear()

        var count = 0
        var minRowIndex = 1
        var maxRowIndex: Int
        val totalSectionCount = getTotalSectionCount()

        for (section in 0 until totalSectionCount) {
            val rowCount = getSectionItemCount(section)
            maxRowIndex = minRowIndex + rowCount
            sectionPositions.add(section, minRowIndex - 1)
            sectionInfos.add(section, DefaultSurveySectionInfo(section, minRowIndex, maxRowIndex, rowCount))

            count += rowCount

            minRowIndex = maxRowIndex + 1
        }

        return count + totalSectionCount
    }

    private fun getTotalSectionCount(): Int {
        return sections.count()
    }

    private fun getSectionItemCount(section: Int): Int {
        return sections[section].items.count()
    }
}

